arch/i386/mcount-support.c

```diff
@@ -0,0 +1,352 @@
+		memcpy(target_addr, trampoline, sizeof(trampoline));
+	}
+
+	pr_dbg2("real address to jump: %#lx\n", plt_addr);
+	memcpy(trampoline_buf + (idx * sizeof(trampoline)),
+	       &plt_addr, sizeof(plt_addr));
+
+	mprotect(trampoline_buf, trampoline_size, PROT_READ | PROT_EXEC);
+	return 0;
+}
+
+unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
+{
+	struct sym *sym;
+
+	sym = &pd->dsymtab.sym[idx];
+	return sym->addr + ARCH_PLTHOOK_ADDR_OFFSET;
+}
+
+
+/*
+	For 16-byte stack-alignment, 
+	the main function stores the return address in its stack scope at prologue.
+	When the time comes for the main function to return,
+	1. restore the saved return address from stack.
+	2. After cleaning up the stack.
+	3. Put the return address at the top of the stack and return.
+	4. will be returned.
+
+	080485f8 <main>:
+	80485f8: 8d 4c 24 04           lea    0x4(%esp),%ecx
+	80485fc: 83 e4 f0              and    $0xfffffff0,%esp
+	80485ff: ff 71 fc              pushl  -0x4(%ecx)
+	8048602: 55                    push   %ebp
+	8048603: 89 e5                 mov    %esp,%ebp
+	8048605: 51                    push   %ecx
+	8048606: 83 ec 14              sub    $0x14,%esp
+	8048609: e8 02 fe ff ff        call   8048410 <mcount@plt>
+
+	... ... 
+
+	8048645: 8b 4d fc              mov    -0x4(%ebp),%ecx
+	8048648: c9                    leave
+	8048649: 8d 61 fc              lea    -0x4(%ecx),%esp
+	804864c: c3                    ret
+
+	So, in this case. The return address we want to replace with 
+	mcount_exit is in the stack scope of the main function. 
+	Non a parent located. 
+
+	we search stack for that address. 
+	we will look for it.
+	we will find it, and we will replace it. 
+	GOOD LUCK!
+*/
+unsigned long *mcount_arch_parent_location(struct symtabs *symtabs,
+                                            unsigned long *parent_loc, 
+                                            unsigned long child_ip)
+{
+	if (!main_ret_found) {
+		struct sym *parent_sym, *child_sym;
+		char *parent_name, *child_name;
+
+		const char *find_main[] = {
+			"__libc_start_main",
+			"main"
+		};
+		unsigned long ret_addr;
+		unsigned long search_ret_addr;
+
+		ret_addr = *parent_loc;
+		parent_sym = find_symtabs(symtabs, ret_addr);
+		parent_name = symbol_getname(parent_sym, ret_addr);
+		child_sym = find_symtabs(symtabs, child_ip);
+		child_name = symbol_getname(child_sym, child_ip);
+
+		// Assuming that this happens only in main..			
+		if (!(strcmp(find_main[0], parent_name) || strcmp(find_main[1], child_name))) {
+			ret_addr = *parent_loc;
+			for (int i = 1; i < 5; i++ ) {
+				search_ret_addr = *(unsigned long *)(parent_loc + i);
+				if (search_ret_addr == ret_addr) {
+					parent_loc = parent_loc+i;
+					main_ret_found = true;
+				}
+			}
+		}
+		
+		// if we couldn't found correct position of return address,
+		// maybe this approach is not available anymore.
+		if (!main_ret_found) {
+			pr_dbg2("cannot find ret address of main\n");
```

Comment:
then I don't want to print it repeatedly.  What about force setting to true here?

Reply:
As I said, my comment assumed that it already returned for non-main functions.  Otherwise if it runs any other (e.g some initialization) function before main, this logic won't work correctly.